### PR TITLE
New METAL DETECTORS for LCZ

### PR DESCRIPTION
**Pull request by aidan-bowman**
## About the Pull Request

Technically contraband detectors, but most people call them by the same name.
Mobs (and items, so you can't just throw your backpack across) that pass through are scanned for contraband status, right now it's basically just guns, knives, shards, and tools, but it could be expanded decently easily.
There are some in LCZ by default (blocking off mining, the canteen, and the kitchen), although engineers can easily make more.
Also adds an aluminum box that can pass undetected through said scanner (including the contents inside) so that'll be nice.

Has wires system for fun sabotage gameplay; turn off power, the radio system, or make it think everything is contraband and frame that guy you don't like.

## Why It's Good For The Game

LCZ guards need something to help them control D-class even at lowpop, and D-class need obstacles to push them towards sneakier gameplay.

## Future Improvements

Wiring system looked completely fucked and needs to be fixed sometime.

## Changelog

:cl:
add: Adds new contraband detectors! Found in LCZ D-block next to mining, and constructable by engineers, these scanners search personnel that walk through it for contraband (guns, tools, knives, etc). There are also ways for clever D-class to bypass or even abuse them -- check the codex entry for more information.
/:cl:

### CODE REVIEW

File: code/game/atoms_movable.dm [hunk 1]
```diff
@@ -125,6 +125,8 @@
 			for (var/datum/light_source/L in light_sources)
 				L.source_atom.update_light()
 
+		SEND_SIGNAL(loc, COMSIG_TURF_CROSSED, src)
+
```
Comment (Tsurupeta) [inline code/game/atoms_movable.dm]: are you retarded?

**Review by Tsurupeta — CHANGES_REQUESTED** [review]

File: code/game/atoms_movable.dm (hunk 1 above)
Comment (aidan-bowman) [reply code/game/atoms_movable.dm]: ????

Comment (Tsurupeta) [reply code/game/atoms_movable.dm]: Move is not Crossed, there's entire proc called CROSSED and also CROSS

Comment (cheesePizza2) [reply code/game/atoms_movable.dm]: renamed

File: code/game/machinery/contraband_detector.dm [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+	req_access = list(list(ACCESS_SECURITY_LVL2))
+	var/id_locked = TRUE
+	var/wrenching = FALSE
+	var/obj/item/device/radio/intercom/announce
+
+	wires = /datum/wires/contraband_detector
+	var/power_wire_cut = FALSE				// if cut... no power. also electrifies cutter, so there's some danger here
+	var/identifier_wire_cut = FALSE			// if cut, registers nothing as contraband. if pulsed, registers everything
+	var/identifier_wire_pulsed_until = 0	// world time value, since it's only checked in 1 place we don't remove it when time has passed (more optimal than checking every tick)
+	var/radio_wire_cut = FALSE				// if cut, removes ability to radio
```
Comment (EgorDinamit) [inline code/game/machinery/contraband_detector.dm]: You have wires for that, why make it into variables?

File: code/game/machinery/contraband_detector.dm [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+	announce.set_frequency(SEC_LCZ_FREQ)
+
+/obj/machinery/contraband_detector/set_broken(new_state)
+	. = ..()
+	if(new_state)
+		announce_tampering()
+
+/obj/machinery/contraband_detector/ex_act(severity)
+	. = ..()
+	if(prob(100 - (severity * 20)))
```
Comment (EgorDinamit) [inline code/game/machinery/contraband_detector.dm:42]: Maybe qdel at devastation range/severity?

File: code/game/machinery/contraband_detector.dm [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+		src.set_broken(TRUE)
```
Comment (EgorDinamit) [inline code/game/machinery/contraband_detector.dm]: src is always implied, you don't have to put it here.

File: code/game/machinery/contraband_detector.dm (hunk 3 above)
Comment (aidan-bowman) [reply code/game/machinery/contraband_detector.dm:42]: That already happens in the ..()

**Comment by aidan-bowman** [comment]
welp wires are broken, i can't fucking figure out why though. will take a look at em later

**Comment by Tsurupeta** [comment]
I'll TM it

**Comment by TheDarkElites** [comment]
Ill throw this in for tming tmrw but I'd like for @Tsurupeta to look over it and give his approval before it gets fully merged.

**Comment by TheDarkElites** [comment]
image
this needs a cooldown of some sort

**Comment by TheDarkElites** [comment]
also please make it so you can jump to the detector from the message

File: code/game/machinery/contraband_detector.dm [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+
+/obj/machinery/contraband_detector_construct
+	name = "contraband detector shell"
+	desc = "You can anchor the shell down with a wrench, or dismantle it with a crowbar."
+	icon = 'icons/obj/machines/contraband_detector.dmi'
+	icon_state = "shell"
+	density = TRUE
+	var/build_step = 0										// the current step in the building process
+
+/obj/machinery/contraband_detector_construct/attackby(obj/item/I, mob/user)
```
Comment (TheDarkElites) [inline code/game/machinery/contraband_detector.dm]: This should use the /obj/machinery/constructable_frame system as any other machine.

**Review by TheDarkElites — CHANGES_REQUESTED** [review]
See Comments

**Comment by 77seven77seven** [comment]
detection system works on borgs, no need to flood the chat more

File: code/game/machinery/contraband_detector.dm (hunk 5 above)
Comment (aidan-bowman) [reply code/game/machinery/contraband_detector.dm]: done

File: code/game/machinery/contraband_detector.dm [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+				update_icon()
+				to_chat(user, SPAN_NOTICE("You secure the exterior bolts on the contraband detector."))
+				RegisterSignal(loc, COMSIG_ENTERED, .proc/detect_contraband)
+			else if(anchored)
+				playsound(loc, 'sound/items/Ratchet.ogg', 100, 1)
+				anchored = FALSE
+				to_chat(user, SPAN_NOTICE("You unsecure the exterior bolts on the contraband detector."))
+				update_icon()
+				UnregisterSignal(loc, COMSIG_ENTERED)
+		wrenching = FALSE
```
Comment (TheDarkElites) [inline code/game/machinery/contraband_detector.dm]: This is handled by obj/machinery

If you need to override for checks just do the check and then return ..()

File: code/game/machinery/contraband_detector.dm [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+	if(isScrewdriver(I))
+		if(id_locked && !panel_open)	// you can close it when ID locked, just not open it
+			to_chat(user, SPAN_NOTICE("\The [src]'s panel is locked!"))
+		else
+			panel_open = !panel_open
+			user.visible_message(SPAN_WARNING("[user] screws the camera's panel [panel_open ? "open" : "closed"]!"),
+			SPAN_NOTICE("You screw the camera's panel [panel_open ? "open" : "closed"]."))
+			playsound(src.loc, 'sound/items/Screwdriver.ogg', 50, 1)
+	if((isWirecutter(I) || isMultitool(I)) && panel_open)
+		return wires.Interact(user)
```
Comment (TheDarkElites) [inline code/game/machinery/contraband_detector.dm]: Same with this, dont override just do any special checks and use obj/machinery's unarmedAttack override

**Review by TheDarkElites — CHANGES_REQUESTED** [review]

**Review by TheDarkElites — DISMISSED** [review]
Passed TM and it looks good to me.

File: code/datums/observation/moved.dm [hunk 8]
```diff
@@ -28,6 +28,11 @@ GLOBAL_DATUM_INIT(moved_event, /decl/observ/moved, new)
 /atom/Entered(atom/movable/am, atom/old_loc)
 	. = ..()
 	GLOB.moved_event.raise_event(am, old_loc, am.loc)
+	SEND_SIGNAL(src, COMSIG_ENTERED, am, old_loc)
+
+/atom/Exited(atom/movable/exitee, atom/new_loc)
+	. = ..()
+	SEND_SIGNAL(src, COMSIG_EXITED, exitee, new_loc)
```
Comment (Tsurupeta) [inline code/datums/observation/moved.dm]: We already have this code in `code/datums/observation/entered.dm` and `exited.dm`

**Review by Tsurupeta — DISMISSED** [review]
^

**Review by TheDarkElites — APPROVED** [review]

**Pull request merged**